README.md
# Computer-Science-And-Technology
换行使用空格+空格+回车  
Study Notes for the Computer Science And Technology  
include the java source code  
include the BIO NIO AIO  
include the Annotation of java etc


history.txt
 8915  2021-03-18 16:22:54 root cd /data/wwwroot/
 8916  2021-03-18 16:22:54 root sl
 8917  2021-03-18 16:22:55 root ls
 8918  2021-03-18 16:23:03 root cd qpyx-parent/qpyx-app/
 8919  2021-03-18 16:23:04 root ls
 8920  2021-03-18 16:23:06 root cd target/
 8921  2021-03-18 16:23:07 root ls
 8922  2021-03-22 11:40:57 root cd zero
 8923  2021-03-22 11:40:58 root ls
 8924  2021-03-22 11:41:44 root yum install -y yum-utils device-mapper-persistent-data lvm2
 8925  2021-03-22 11:42:59 root yum-config-manager --add-repo https://download.docker.com/linux/centos/docker-ce.repo
 8926  2021-03-22 11:51:37 root cd zero
 8927  2021-03-22 11:51:38 root ls
 8928  2021-03-22 11:51:52 root yum-config-manager --enable docker-ce-edge
 8929  2021-03-22 11:52:01 root yum install -y docker-ce
 8930  2021-03-22 11:53:02 root docker version
 8931  2021-03-22 11:53:14 root systemctl docker
 8932  2021-03-22 11:53:31 root systemctl start docker
 8933  2021-03-22 11:54:04 root docker version
 8934  2021-03-22 11:54:48 root ls
 8935  2021-03-22 11:56:00 root java version
 8936  2021-03-22 11:56:09 root java -version
 8937  2021-03-22 13:34:59 root docker version
 8938  2021-03-22 13:39:58 root docker images
 8939  2021-03-22 13:40:33 root docker ps
 8940  2021-03-22 13:42:11 root ls
 8941  2021-03-22 13:42:27 root vi Dockerfile
 8942  2021-03-22 13:50:13 root ls
 8943  2021-03-22 13:52:26 root cat Dockerfile
 8944  2021-03-22 13:56:15 root ifconfig
 8945  2021-03-22 14:04:40 root ls
 8946  2021-03-22 14:04:51 root vi Dockerfile
 8947  2021-03-22 14:06:17 root docker built -t DockerOne .
 8948  2021-03-22 14:07:39 root docker command --help
 8949  2021-03-22 14:10:25 root ls
 8950  2021-03-22 14:10:30 root docker built -t DockerOne .
 8951  2021-03-22 14:10:34 root ls
 8952  2021-03-22 14:15:23 root docker build -t DockerOne .
 8953  2021-03-22 14:15:52 root docker build -t BlockOne .
 8954  2021-03-22 14:16:19 root ls
 8955  2021-03-22 14:16:35 root vi Dockerfile
 8956  2021-03-22 14:18:12 root docker build -t blockOne .
 8957  2021-03-22 14:18:38 root vi Dockerfile
 8958  2021-03-22 14:19:15 root docker build -t blockOne .
 8959  2021-03-22 14:19:34 root docker build -t blockone .
 8960  2021-03-22 14:21:55 root docker images
 8961  2021-03-22 14:22:16 root docker run -it blockone
 8962  2021-03-22 14:23:10 root ls
 8963  2021-03-22 14:23:15 root cd zero
 8964  2021-03-22 14:23:16 root ls
 8965  2021-03-22 14:23:24 root docker ps
 8966  2021-03-22 14:23:58 root docker inspect 92a3e
 8967  2021-03-22 14:26:12 root curl 92a3e858723a:8080/scan
 8968  2021-03-22 14:27:16 root curl 92a3e:8080/scan
 8969  2021-03-22 14:27:53 root docker inspect 92a3e858723a
 8970  2021-03-22 14:28:02 root curl 92a3e858723a:8080/scan
 8971  2021-03-22 14:29:54 root docker ps
 8972  2021-03-22 14:30:52 root curl 172.17.0.2:8080/scan
 8973  2021-03-22 14:31:10 root curl 92a3e858723a:8080/scan
 8974  2021-03-22 14:32:24 root curl 172.17.0.2:8080/create
 8975  2021-03-22 14:36:27 root vi Dockerfile
 8976  2021-03-22 14:40:04 root history
 8977  2021-03-22 14:49:48 root curl 92a3e858723a:8080/scan
 8978  2021-03-22 14:49:56 root curl 172.17.0.2:8080/create
 8979  2021-03-22 16:45:04 root ls
 8980  2021-03-22 16:51:04 root cd node
 8981  2021-03-22 16:51:13 root vi Dockerfile
 8982  2021-03-22 16:57:10 root docker build -t blocktwo .
 8983  2021-03-22 16:57:17 root docker ps
 8984  2021-03-22 16:58:40 root docker images
 8985  2021-03-22 16:59:07 root docker run -it blocktwo
 8986  2021-03-22 16:59:56 root ls
 8987  2021-03-22 17:00:02 root cd zero/node
 8988  2021-03-22 17:00:02 root ls
 8989  2021-03-22 17:00:18 root docker ps
 8990  2021-03-22 17:00:38 root docker inspect 82603891b4e2
 8991  2021-03-22 17:01:50 root curl 82603891b4e2:8090/scan
 8992  2021-03-22 17:05:10 root curl 127.0.0.3:8090/scan
 8993  2021-03-22 17:05:25 root curl 127.17.0.3:8090/scan
 8994  2021-03-22 17:07:00 root curl 127.17.0.3:8090/mine
 8995  2021-03-22 17:07:33 root curl 172.17.0.3:8090/min
 8996  2021-03-22 17:07:41 root curl 172.17.0.3:8090/mine
 8997  2021-03-22 17:21:26 root curl 172.17.0.3:8090/create
 8998  2021-03-22 17:21:36 root curl 172.17.0.3:8090/mine
 8999  2021-03-22 11:16:39 root ll
 9000  2021-03-22 11:16:44 root cd /mnt/
 9001  2021-03-22 11:16:44 root ll
 9002  2021-03-22 11:16:50 root cd /data/wwwroot/
 9003  2021-03-22 11:16:51 root ll
 9004  2021-03-23 10:32:14 root ncv
 9005  2021-03-23 10:32:19 root ll
 9006  2021-03-23 10:32:26 root cat testadmin.qingpinyouxuan.com.conf 
 9007  2021-03-23 11:10:08 root history | grep zabbix
 9008  2021-03-23 11:10:37 root /etc/init.d/zabbix_server restart
 9009  2021-03-23 11:10:42 root /etc/init.d/zabbix_server status
 9010  2021-03-23 11:10:48 root ss -ntl
 9011  2021-03-23 11:10:52 root ncv
 9012  2021-03-23 11:10:53 root ls
 9013  2021-03-23 11:11:01 root cat testzabbix.qingpinyouxuan.com.conf
 9014  2021-03-23 11:25:58 root ncv
 9015  2021-03-23 11:26:01 root cd /data/wwwlogs/
 9016  2021-03-23 11:26:02 root ls
 9017  2021-03-23 11:26:05 root pwd
 9018  2021-03-23 11:26:19 root ncv
 9019  2021-03-23 11:26:20 root cd ..
 9020  2021-03-23 11:26:20 root ls
 9021  2021-03-23 11:26:23 root vim nginx.conf
 9022  2021-03-23 11:26:40 root cd /data/wwwlogs/
 9023  2021-03-23 11:26:50 root tail -n1 testapi.qingpinyouxuan.com_nginx.log
 9024  2021-03-23 11:31:16 root cd /data/wwwroot/
 9025  2021-03-23 11:31:17 root ls
 9026  2021-03-23 10:10:53 root ls
 9027  2021-03-23 10:10:56 root cd zero
 9028  2021-03-23 11:42:26 root cd ..
 9029  2021-03-23 11:42:31 root ls
 9030  2021-03-23 11:42:37 root cd ..
 9031  2021-03-23 11:42:39 root ls
 9032  2021-03-23 11:42:42 root cd data
 9033  2021-03-23 11:42:44 root ls
 9034  2021-03-23 11:42:48 root cd wwwlgos
 9035  2021-03-23 11:45:48 root wwwlogs
 9036  2021-03-23 11:45:53 root cd wwwlogs
 9037  2021-03-23 11:45:56 root ls
 9038  2021-03-23 11:46:07 root cd print-app
 9039  2021-03-23 11:46:08 root ls
 9040  2021-03-23 11:46:30 root tail -f all.log --line 1000

DataStructure/String/main.c
#include <stdio.h>
#include <stdbool.h>


#define MAXLEN 255 //预定最大的串长度255。

//静态数组定义串，静态数组实现定长顺序存储，当有关静态数组实现的代码执行成功后，其所占的空间会自动被回收。
typedef  struct{
    char ch[MAXLEN]; //每个分量存储一个字符串。
    int length;  //串的实际长度。
}SString;

//动态数组定义串,动态数组实现是基于堆分配存储，当有关动态数组的代码执行完之后，堆空间需要通过free函数手动回收。
typedef struct{
    char *ch;
    int length;
}HString;

/*
 * HString S;
 * S.ch = (char *) malloc(MAXLEN * sizeof(char));
 * S.length = 0;
 * */

//求子串,用Sub返回串S的第pos个字符起长度为len的子串
bool SubString(SString *Sub,SString S, int pos, int len){

    //子串范围越界
    if(pos+len-1 > S.length){
        return false;
    }

    for(int i = pos; i < pos + len; i++){
        Sub->ch[i-pos+1] = S.ch[i];
        Sub->length = len;
    }
    return true;
}

//子串与主串的比较操作。若S>T,则返回值 > 0; 若S = T, 则返回值 = 0； 若S < T, 则返回值 < 0
int StrCompare(SString S,SString T) {
    for (int i = 1; i <= S.length && i <= T.length; i++) {
        if(S.ch[i] != T.ch[i]) {
            return S.ch[i] - T.ch[i];
        }
    }
    //扫描过的所有字符都相同，则长度长的串更大
    return S.length - T.length;
}

//定位操作。若主串S中存在与串T值相同的子串，则返回它在主串S中第一次出现的位置，否则函数值为0.
int Index(SString S, SString T) {
    int i = 1, n = S.length, m = T.length;
    SString *sub;
    while(i <= n-m+1) {
        SubString(sub,S,i,m);
        if(StrCompare(*sub,T) != 0) ++i;
        else return i; //返回子串在主串的位置
    }
    return 0;   //S中不存在与T相同的子串
}

int f(char *s) {
    int n = 0, i;
    for(i = 0; s[i] != '\0'; i++) {
        //简单十六进制数0~15为：0，1，2，3，4，5，6，7，8，9，a，b，c，d，e，f
        //0~15对应的ASCII表为：48，49，50，51，52，53，54，55，56，97，98，99，100，101，102
        //'a'到'f'的ASCII码是97到102
        //s[i] >= 97在ASCII码表中为a及以后的字符
        //s[i] - 87表示为，如果十六进制数中某位为a,b,c,d,e,f中的某一位，则对应10进制中的10，11，12，13，14，15
        //s[i] - 48表示为，十六进制数中某位为0~9，则满足该规则
        n = n*16 + (s[i] >= 97 ? s[i] - 87:s[i] - 48);
    }
    return n;
}

int main() {
    char s[10];
    scanf("%s",s);
    printf("%d\n",f(s));
    return 0;
}